### ISSUE TITLE
Nuget can't find `msbuild.exe` on Mono?

### ISSUE THREAD

**Issue by ethomson** [NuGet/Home#3085]
[https://github.com/libgit2/LibGit2Sharp](LibGit2Sharp) is trying to [get fancy with the new .NET Core](https://github.com/libgit2/libgit2sharp/pull/1318).  It seems like things are working with the Nuget 3.5.0 beta on Windows with the proper .NET CLR.  However, `nuget restore` is not working for us on Mono:

```
% mono Lib/NuGet/Nuget.exe restore
MSBuild auto-detection: using msbuild version '4.0' from '/Library/Frameworks/Mono.framework/Versions/4.4.1/lib/mono/4.5'.
MsBuild.exe does not exist at '/Library/Frameworks/Mono.framework/Versions/4.4.1/lib/mono/4.5/msbuild.exe'.
```

There is, of course, no `msbuild.exe` on Mono.  :(

Not sure if this is not supported at all (in which case a more helpful error message would be nice) or if this should actually be looking for `xbuild.exe` in the same location on Mono.

**Comment by rrelyea** [comment]
Are you using 3.5.0-beta2 of NuGet.exe?
Which build of NuGet.exe last ran properly for you on Mono?
What project are you trying restore? LibGit2Sharp?

**Comment by ethomson** [comment]
> Are you using 3.5.0-beta2 of NuGet.exe?

Yes:

```
% mono Lib/NuGet/NuGet.exe help
NuGet Version: 3.5.0.1520
```

> Which build of NuGet.exe last ran properly for you on Mono?

AFAIK, this has never worked with our new .NET Core compatible project structure.  We are (successfully) building the old-style project structure using `NuGet Version: 2.8.60318.667`.

> What project are you trying restore? LibGit2Sharp?

Yes - we're trying to run package restore to build LibGit2Sharp.

**Comment by MichalTelus** [comment]
Hi! I have the same problem on version 3.4.4.1321. I want use it with Bamboo on OS X, is other way to restore packages? Or is some beta version with fixed this bug?

**Comment by radical** [comment]
@ethomson Mono has added a shell script called `msbuild` which executes the open sourced `msbuild` and _not_ `xbuild.exe`. It runs that from `/Library/Frameworks/Mono.framework/Versions/4.4.1/lib/mono/msbuild/14.1/bin/MSBuild.exe` .

**Comment by ethomson** [comment]
@radical That does not help me, though, unless I can inform nuget of this location.

**Comment by radical** [comment]
Maybe [this commit](https://github.com/NuGet/NuGet.Client/commit/f2f89d45fb2f330a20cd79c2d7654f0cc50f7e4a) broke it? It changes the logic to look for `msbuild.exe` specifically now. On Mono, it can should try to use `xbuild`, that will be available in `$PATH`. Or if they wanted to use the _preview_ msbuild on Mono, then look for `msbuild` in `$PATH`.

**Comment by kpranavk** [comment]
What's the latest on this thread. I am hitting the exact same error. If that commit broke it, what is the suggestion you are making? Is msbuild or xbuild the right thing to use. I am on Xamarin.iOS

**Comment by yishaigalatzer** [comment]
Can you try passing -msbuildversion flag (that avoids the detection phase)

**Comment by pjcollins** [comment]
When running with mono, `4` appears to be the only recognized msbuild version. This failure still manifests in the same way mentioned above when also passing `-msbuildversion 4`.

**Comment by yishaigalatzer** [comment]
@pjcollins thanks. We will take a look and find a way around this,

@ethomson any reason you are not using `dotnet restore`?

**Comment by ravensorb** [comment]
Any update on this? Maybe a build eta?

**Comment by yishaigalatzer** [comment]
We don't have an update on this at the moment, we never planned to support .net core builds on mono with nuget.exe.

We will however look at:
1. Can we run msbuild for restoring a solution on mono for packages.config scenarios (3.5 +)
2. Can we nuget.exe restore for a solution for the final .net core mode. (3.6 +)

At this point in time if you want to build .net core on linux use `dotnet restore`.

**Comment by jtattermusch** [comment]
@yishaigalatzer, is this the issue to track for "1. Can we run msbuild for restoring a solution on mono for packages.config scenarios (3.5 +)". Is there an ETA for the 3.5 release?
For our project (gRPC) this is important because the user experience is really bad when developers cannot run `nuget restore` for our solutions on mono. We are in a weird state where one needs at least nuget 2.12, but that one is not officially supported and nuget 3.4 is broken.

**Comment by ravensorb** [comment]
I agree - I to have several frameworks that can be compiled on mono and are still using package.config - there needs to be a way to restore those packages in an easier manner then restoring to a very old version of nuget :)

**Comment by yishaigalatzer** [comment]
Yes, we are looking at a fix for this right now. The range at the moment is within weeks for an official release. The mono team did release their own version of 2.12 (though we do not support it), does that unblocks you?

**Comment by akamud** [comment]
Is there a workaround for this issue while we don't get an official release?

**Comment by zhili1208** [comment]
@akamud I don't think there is a workaround for this, the issue is NuGet hardcode "msbuild.exe" in code,
I made a fix to fallback to xbuild when msbuild doesn't exist, the fix will be in 3.5-rtm release

**Comment by zhili1208** [comment]
https://github.com/NuGet/NuGet.Client/commit/df9d2eea5930f3ebb74f36387810198550b3509a
